docs/security-self-assessment.md

```diff
@@ -0,0 +1,268 @@
+
+**Database Server**
+
+The database server stores relationship tuples, authorization models and a changelog. OpenFGA currently supports Postgres and MySQL, and it can be extended to support other databases.
+
+Database credentials can be provided as a parameter to the OpenFGA server, via environment variables, or a configuration file.
+
+**CLI/API Clients**
+
+The CLI/API clients are used to make API requests to the OpenFGA server. This includes functionality such as creating and querying relationship tuples, and updating the authorization model.
+
+They authenticate using any of the mechanisms supported by the server (no authentication, shared key, OAuth client credentials).
+
+If the credentials were compromised, the attacker would be able to perform any action granted to those credentials. This would include: 
+
+* Changing the authorization model.
+* Adding tuples to give the attacker access to a resource.
+* Removing tuples to prevent a user from accessing a resource.
+* Querying authorization data, giving the attacker access to privileged information.
+
+To achieve this the attacker would need to be able to connect directly to the OpenFGA service, which should not be accessible from the public internet.
+
+OpenFGA provides versioned authorization models, and applications should point to a specific version. This implies that even if the attacker updates the authorization model, it should not have any effect unless they can also update the application configuration to make it point to the new version.
+
+### Actions
+
+**Invoking the OpenFGA APIs**
+
+Every time a server endpoint is invoked, OpenFGA validates that:
+
+  - The credentials provided in the API call match the ones configured in the server.
+
+  - The payload of the API call:
+    - Matches the Protobuf API definitions.
+    - Validates the parameters have the proper structure, e.g. users need to be written this way 'user:<userid>'
+  
+**Writing an Authorization Model**
+
+OpenFGA Authorization Models should be Directed Acyclic Graphs. OpenFGA validates that the model does not have cycles.
+
+When a model is written a new version is created. The application needs to be configured to use the new version when needed.
+
+**Calling the Authorization Query endpoints**
+
+When the [/check](https://openfga.dev/api/service#/Relationship%20Queries/Check) and [/list-objects](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects) endpoints are called, OpenFGA needs to traverse the graph defined by the model and instated by the tuples. The graph can be very wide and deep. To protect the service from a DoSS, OpenFGA needs to bound the number of simultaneous paths it explores.
+
+**Upgrading OpenFGA Database Schema**
+
+When new version of OpenFGA are installed, it might imply to run a database migration in the target system. To avoid downtime, migrations need to be carefully written and planned.
```

Comment:
```suggestion
When new version of OpenFGA are installed, it might imply running a database migration in the target system. To avoid downtime, migrations need to be carefully written and planned.
```